### PR TITLE
[Dashboard] Add task detail page with logs

### PR DESCRIPTION
**Pull request by alanwguo**
## Why are these changes needed?
- Add task detail page with logs
- Update actor logs to use state-api actor_id filter instead of filename to fetch logs
- Hide log tabs if there is only 1 tab available.
- Update the driver logs in the jobs page to show explanation why logs are not available.

Screenshot 2023-05-14 at 5 23 32 PM
Screenshot 2023-05-14 at 5 23 14 PM

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by scottsun94** [comment]
Awesome! 2 minor comments:
1. Shall we also add copy button for actor id and node id?
2. RE: "Driver logs are only available when submitting jobs via the Job Submission API or the `ray job submit` CLI command."
- Is it clear that users can also use SDK? Shall we say something like "Job Submission API, SDK or ... CLI..."?
- Shall we add a "Lean more" link at the end which points to https://docs.ray.io/en/latest/cluster/running-applications/job-submission/index.html

**Comment by galenhwang** [comment]
(unrelated to this PR)

nit: The `Node Status` panel in `Cluster status and autoscaler` seems to have rounded corners (inconsistent with other panels) which means some text is hidden.

Screenshot 2023-05-15 at 12 19 48 PM

**Comment by alanwguo** [comment]
> Awesome! 2 minor comments:
> 
> 1. Shall we also add copy button for actor id and node id?
> 2. RE: "Driver logs are only available when submitting jobs via the Job Submission API or the `ray job submit` CLI command."
> 
> * Is it clear that users can also use SDK? Shall we say something like "Job Submission API, SDK or ... CLI..."?
> * Shall we add a "Lean more" link at the end which points to https://docs.ray.io/en/latest/cluster/running-applications/job-submission/index.html

1) yes
2) I can add the url as a message but it won't be easy to make it a clickable link.

**Comment by alanwguo** [comment]
> (unrelated to this PR)
> 
> nit: The `Node Status` panel in `Cluster status and autoscaler` seems to have rounded corners (inconsistent with other panels) which means some text is hidden.
> 
> Screenshot 2023-05-15 at 12 19 48 PM

ill put in this fix

**Comment by alanwguo** [comment]
Copyable links for actor_id and node_id:
Screenshot 2023-05-15 at 12 52 18 PM

Fix overview cards:
Screenshot 2023-05-15 at 12 51 37 PM

**Comment by scottsun94** [comment]
> I can add the url as a message but it won't be easy to make it a clickable link.
Non-clickable url is enough. Thanks!

**Comment by alanwguo** [comment]
Screenshot 2023-05-15 at 1 28 35 PM

File: dashboard/client/src/service/log.ts [hunk 1]
```diff
@@ -50,13 +50,56 @@ export const getLogDetail = async (url: string) => {
...
+  if (
+    nodeId === null ||
+    actorId === null ||
+    taskId === null ||
+    filename === null
+  ) {
+    // Null means data is not ready yet.
+    return null;
+  }
+  const variables = [
```
Comment (galenhwang) [inline dashboard/client/src/service/log.ts:77]: nit: could we use `URLSearchParams` and its `toString()` here? something like

```
export const getStateApiDownloadLogUrl = ({
  ...args
}: StateApiLogInput) => {
  const path = new URLSearchParams();
  const allArgs = {
    ...args,
    lines: -1,
  }
  for (const key in allArgs) {
    const x = allArgs[key];
    if (x === null) {
      return null;
    }
    if (x !== undefined) {
      path.set(ARG_NAME_TO_SEARCH_PARAM_MAP[key], x);
    }
  }

  return `api/v0/logs/file?${path.toString()}`;
};
```

**Review by galenhwang — APPROVED** [review]

File: dashboard/client/src/common/MultiTabLogViewer.tsx [hunk 2]
```diff
@@ -119,23 +125,114 @@ export const MultiTabLogViewer = ({
...
-}: StateApiLogViewerProps) => {
-  const { downloadUrl, log, path, refresh } = useStateApiLogs(nodeId, filename);
+  actorId,
+  suffix,
+}: {
+  height: number;
+} & ActorData) => {
+  const apiData = useStateApiLogs(
+    { actorId, suffix },
+    `actor-log-${actorId}.${suffix}`,
```
[the remark this answers is not in the document]
Comment (alanwguo) [reply dashboard/client/src/common/MultiTabLogViewer.tsx:205]: what do you mean by not used for filtering?

File: dashboard/client/src/components/MetadataSection/MetadataSection.tsx [hunk 3]
```diff
@@ -92,6 +96,28 @@ export const MetadataContentField: React.FC<{
...
+        size="small"
+        className={classes.button}
+      >
+        <RiFileCopyLine />
+      </IconButton>
+    </Tooltip>
+  );
+
   if (content === undefined || "value" in content) {
     return content === undefined || !("link" in content) ? (
```
Comment (chaowanggg) [inline dashboard/client/src/components/MetadataSection/MetadataSection.tsx:122]: nit: Could we simplify the logic here? Like early catch content == undefined


if(content == undefined) return <...>

if ("value" in content) {
    .... 
}
if ("link" in content) {
    .... 
}

File: dashboard/client/src/App.tsx [hunk 4]
```diff
@@ -201,16 +202,23 @@ const App = () => {
                     <Route
                       element={
                         <JobDetailActorDetailWrapper>
-                          <ActorDetailPage />
+                          <ActorDetailLayout />
```
[the remark this answers is not in the document]
Comment (alanwguo) [reply dashboard/client/src/App.tsx:205]: I don't quite understand the question. The ActorDetailPage is a child of this ActorDetailLayout

File: dashboard/client/src/common/MultiTabLogViewer.tsx [hunk 5]
```diff
@@ -59,52 +61,56 @@ export const MultiTabLogViewer = ({
...
-                  setValue(value);
-                }}
-                component={Link}
-                to={otherLogsLink}
-                target="_blank"
-                rel="noopener noreferrer"
-              />
-            )}
-          </Tabs>
+          {(tabs.length > 1 || otherLogsLink) && (
```
[the remark this answers is not in the document]
Comment (alanwguo) [reply dashboard/client/src/common/MultiTabLogViewer.tsx:64]: Screenshot 2023-05-15 at 3 07 02 PM

File: dashboard/client/src/common/MultiTabLogViewer.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (alanwguo) [reply dashboard/client/src/common/MultiTabLogViewer.tsx:205]: isn't that clear by the type definition?
```
type TextData = {
  contents: string;
};
type FileData = {
  nodeId: string | null;
  filename?: string;
};
type ActorData = {
  actorId: string | null;
  suffix: "out" | "err";
};
type TaskData = {
  taskId: string | null;
  suffix: "out" | "err";
};

type LogViewerData = TextData | FileData | ActorData | TaskData;
```

[the remark this answers is not in the document]
Comment (galenhwang) [reply dashboard/client/src/common/MultiTabLogViewer.tsx:205]: I would still add the comment. If someone were to look at the base `StateApiLogInput` type, it's not clear that the fields should really be used as a oneOf

File: dashboard/client/src/App.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (galenhwang) [reply dashboard/client/src/App.tsx:205]: ah I misread the code, the `Layout` is wrapped so ActorDetail and Task are children, makes sense

**Pull request merged**